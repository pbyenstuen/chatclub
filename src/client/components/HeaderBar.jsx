import React from "react";
import { Link } from "react-router-dom";
import { useHistory } from "react-router";
import useSubmit from "./useSubmit";

const HeaderBar = ({ api, user, updateUser }) => {
    const history = useHistory();

    const { handleSubmit: handleLogout, submitting, error } = useSubmit(
        async () => {
            await api.auth.logOut();
        },
        async () => {
            await updateUser();
            history.push("/")
        }
    );

    if (user) {
        return (
            <div id="header-bar">
                <h1>Chatclub</h1>
                <nav>
                    <ul>
                        <li>
                            <Link to="/chatters">CHATTERS</Link>
                        </li>
                        <li>
                            <Link to="/">CHAT</Link>
                        </li>
                        <li>
                            <Link to="/profile">PROFILE</Link>
                        </li>
                    </ul>
                </nav>
                <div id="header-bar-user-info">
                    <h3>{error ? "Logout failed" : user.displayName}</h3>
                    <button disabled={submitting} onClick={handleLogout}>SIGN OUT</button>
                </div>
            </div>
        );
    };

    return (
        <div id="header-bar">
            <h1>Chatclub</h1>
            <nav>
                <ul>
                    <li>
                        <Link to="/signup">SIGN UP</Link>
                    </li>
                    <li>
                        <Link to="/login">SIGN IN</Link>
                    </li>
                </ul>
            </nav>
        </div>
    );
};

export default HeaderBar;
