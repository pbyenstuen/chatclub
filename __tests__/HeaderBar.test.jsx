import React from "react";
import ReactDOM from "react-dom";
import { act, isDOMComponent, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import HeaderBar from "../src/client/components/HeaderBar";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("HeaderBar", () => {
    it("renders unauthenticated view", async () => {
        await mountComponent(<HeaderBar user={undefined} />);

        expect(container.innerHTML).toMatchSnapshot();
        expect(isDOMComponent(container.querySelector("h3"))).toEqual(false);
    });

    it("renders authenticated view", async () => {
        const user = {
            username: "User",
            displayName: "User"
        }

        await mountComponent(<HeaderBar user={user} />);

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h3").textContent).toEqual("User");
    });

    it("updates user on logout", async () => {
        const logOut = jest.fn();
        const updateUser = jest.fn();

        let user = {
            username: "User",
            displayName: "User"
        }
        
        const api = {
            auth: {
                logOut
            }
        }

        await mountComponent(<HeaderBar api={api} user={user} updateUser={updateUser} />);

        expect(container.querySelector("h3").textContent).toEqual("User");
        
        await act(async () => {
            await Simulate.click(container.querySelector("button"));
        });
        
        expect(updateUser).toBeCalled();
    });

    it("can show error message", async () => {
        const user = {
            username: "User",
            displayName: "User"
        }

        const api = {
            auth: {
                logOut: () => {
                    throw new Error;
                }
            }
        }

        await mountComponent(<HeaderBar api={api} user={user} />);

        await act(async () => {
            await Simulate.click(container.querySelector("button"));
        });

        expect(container.querySelector("h3").textContent).toEqual("Logout failed");
    });
});