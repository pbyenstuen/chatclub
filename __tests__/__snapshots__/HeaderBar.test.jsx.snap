// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBar renders authenticated view 1`] = `"<div id=\\"header-bar\\"><h1>Chatclub</h1><nav><ul><li><a href=\\"/chatters\\">CHATTERS</a></li><li><a href=\\"/\\">CHAT</a></li><li><a href=\\"/profile\\">PROFILE</a></li></ul></nav><div id=\\"header-bar-user-info\\"><h3>User</h3><button>SIGN OUT</button></div></div>"`;

exports[`HeaderBar renders unauthenticated view 1`] = `"<div id=\\"header-bar\\"><h1>Chatclub</h1><nav><ul><li><a href=\\"/signup\\">SIGN UP</a></li><li><a href=\\"/login\\">SIGN IN</a></li></ul></nav></div>"`;
